dll/win32/shlwapi/autocomp.c

```diff
@@ -0,0 +1,707 @@
+    return S_OK;
+}
+
+static AC_EnumString *AC_EnumString_Construct(SIZE_T capacity);
+
+static HRESULT STDMETHODCALLTYPE
+AC_EnumString_Clone(IEnumString* This, IEnumString **ppenum)
+{
+    AC_EnumString *this, *cloned;
+    SIZE_T i, count;
+    BSTR *pstrs;
+
+    if (!ppenum)
+        return E_POINTER;
+
+    this = (AC_EnumString *)This;
+    cloned = AC_EnumString_Construct(this->m_cstrs);
+    if (!cloned)
+        return E_OUTOFMEMORY;
+
+    count = this->m_cstrs;
+    pstrs = this->m_pstrs;
+    for (i = 0; i < count; ++i)
+    {
+        if (!AC_EnumString_AddBStrNoGrow(cloned, pstrs[i]))
+        {
+            AC_EnumString_Destruct(cloned);
+            return E_FAIL;
+        }
+    }
+
+    cloned->m_capacity = cloned->m_cstrs = count;
+    cloned->m_hwndEdit = this->m_hwndEdit;
+    cloned->m_dwSHACF_ = this->m_dwSHACF_;
+
+    *ppenum = (IEnumString *)cloned;
+    return S_OK;
+}
+
+#define IS_IGNORED_DOTS(sz) ( \
+    sz[0] == L'.' && (sz[1] == 0 || (sz[1] == L'.' && sz[2] == 0)) \
+)
+
+/* directories only */
+static void
+AC_DoDir0(AC_EnumString *pES, LPCWSTR pszDir)
+{
+    LPWSTR pch;
+    WCHAR szPath[MAX_PATH];
+    HANDLE hFind;
+    WIN32_FIND_DATAW find;
+
+    StringCbCopyW(szPath, sizeof(szPath), pszDir);
+    if (!PathAppendW(szPath, L"*"))
+        return;
+
+    pch = PathFindFileNameW(szPath);
+    assert(pch);
+
+    hFind = FindFirstFileW(szPath, &find);
+    if (hFind != INVALID_HANDLE_VALUE)
+    {
+        do
+        {
+            if (IS_IGNORED_DOTS(find.cFileName))
+                continue;
+
+            if (find.dwFileAttributes & FILE_ATTRIBUTE_HIDDEN)
+                continue;
+            if (!(find.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY))
+                continue;
+
+            *pch = UNICODE_NULL;
+            if (PathAppendW(szPath, find.cFileName))
+                AC_EnumString_AddString(pES, szPath);
+        } while (FindNextFileW(hFind, &find));
+    }
+}
+
+/* all filesystem objects */
+static void
+AC_DoDir1(AC_EnumString *pES, LPCWSTR pszDir)
+{
+    LPWSTR pch;
+    WCHAR szPath[MAX_PATH];
+    HANDLE hFind;
+    WIN32_FIND_DATAW find;
+
+    StringCbCopyW(szPath, sizeof(szPath), pszDir);
+    if (!PathAppendW(szPath, L"*"))
+        return;
+
+    pch = PathFindFileNameW(szPath);
+    assert(pch);
+
+    hFind = FindFirstFileW(szPath, &find);
+    if (hFind != INVALID_HANDLE_VALUE)
+    {
+        do
+        {
+            if (IS_IGNORED_DOTS(find.cFileName))
+                continue;
+
+            if (find.dwFileAttributes & FILE_ATTRIBUTE_HIDDEN)
+                continue;
+
+            *pch = UNICODE_NULL;
+            if (PathAppendW(szPath, find.cFileName))
+                AC_EnumString_AddString(pES, szPath);
+        } while (FindNextFileW(hFind, &find));
+    }
+}
+
+static inline void
+AC_DoDir(AC_EnumString *pES, LPCWSTR pszDir, BOOL bDirOnly)
+{
+    if (bDirOnly)
+        AC_DoDir0(pES, pszDir);
+    else
+        AC_DoDir1(pES, pszDir);
```

Comment:
You can merge both functions `AC_DoDir0` and `AC_DoDir1` together back into `AC_DoDir` by using the very same code as `AC_DoDir0`, but replacing the check on the `FILE_ATTRIBUTE_DIRECTORY` flag by:
```c
    if (bDirOnly && !(find.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY))
        continue;
```